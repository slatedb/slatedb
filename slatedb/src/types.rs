use bytes::Bytes;

/// Represents a key-value pair known not to be a tombstone.
#[non_exhaustive]
#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct KeyValue {
    pub key: Bytes,
    pub value: Bytes,
}

impl<K, V> From<(&K, &V)> for KeyValue
where
    K: AsRef<[u8]>,
    V: AsRef<[u8]>,
{
    fn from(record: (&K, &V)) -> Self {
        let key = Bytes::copy_from_slice(record.0.as_ref());
        let value = Bytes::copy_from_slice(record.1.as_ref());
        KeyValue { key, value }
    }
}

/// Represents a key-value pair that may be a tombstone.
#[derive(Debug, Clone, PartialEq)]
pub(crate) struct RowEntry {
    pub(crate) key: Bytes,
    pub(crate) value: ValueDeletable,
    pub(crate) seq: u64,
    pub(crate) create_ts: Option<i64>,
    pub(crate) expire_ts: Option<i64>,
}

impl RowEntry {
    pub(crate) fn new(
        key: Bytes,
        value: ValueDeletable,
        seq: u64,
        create_ts: Option<i64>,
        expire_ts: Option<i64>,
    ) -> Self {
        Self {
            key,
            value,
            seq,
            create_ts,
            expire_ts,
        }
    }

    pub fn estimated_size(&self) -> usize {
        let mut size = self.key.len() + self.value.len();
        // Add size for sequence number
        size += std::mem::size_of::<u64>();
        // Add size for timestamps
        if self.create_ts.is_some() {
            size += std::mem::size_of::<i64>();
        }
        if self.expire_ts.is_some() {
            size += std::mem::size_of::<i64>();
        }
        size
    }

    #[cfg(test)]
    pub(crate) fn new_value(key: &[u8], value: &[u8], seq: u64) -> Self {
        Self {
            key: Bytes::copy_from_slice(key),
            value: ValueDeletable::Value(Bytes::copy_from_slice(value)),
            seq,
            create_ts: None,
            expire_ts: None,
        }
    }

    #[cfg(test)]
    pub(crate) fn new_merge(key: &[u8], value: &[u8], seq: u64) -> Self {
        Self {
            key: Bytes::copy_from_slice(key),
            value: ValueDeletable::Merge(Bytes::copy_from_slice(value)),
            seq,
            create_ts: None,
            expire_ts: None,
        }
    }

    #[cfg(test)]
    pub(crate) fn new_tombstone(key: &[u8], seq: u64) -> Self {
        Self {
            key: Bytes::copy_from_slice(key),
            value: ValueDeletable::Tombstone,
            seq,
            create_ts: None,
            expire_ts: None,
        }
    }

    #[cfg(test)]
    pub(crate) fn with_create_ts(&self, create_ts: i64) -> Self {
        Self {
            key: self.key.clone(),
            value: self.value.clone(),
            seq: self.seq,
            create_ts: Some(create_ts),
            expire_ts: self.expire_ts,
        }
    }

    #[cfg(test)]
    pub(crate) fn with_expire_ts(&self, expire_ts: i64) -> Self {
        Self {
            key: self.key.clone(),
            value: self.value.clone(),
            seq: self.seq,
            create_ts: self.create_ts,
            expire_ts: Some(expire_ts),
        }
    }
}

/// The metadata associated with a `KeyValueDeletable`
/// TODO: can be removed
#[cfg(test)]
#[derive(Debug, Clone, PartialEq)]
pub(crate) struct RowAttributes {
    pub(crate) ts: Option<i64>,
    pub(crate) expire_ts: Option<i64>,
}

/// Represents a value that may be a tombstone.
/// Equivalent to `Option<Bytes>`, but used internally
/// to prevent type confusion between `None` indicating
/// that a key does not exist, and `Tombstone` indicating
/// that the key exists but has a tombstone value.
#[derive(Debug, Clone, PartialEq)]
pub(crate) enum ValueDeletable {
    Value(Bytes),
    Merge(Bytes),
    Tombstone,
}

impl ValueDeletable {
    pub(crate) fn len(&self) -> usize {
        match self {
            ValueDeletable::Value(v) | ValueDeletable::Merge(v) => v.len(),
            ValueDeletable::Tombstone => 0,
        }
    }

    pub(crate) fn as_bytes(&self) -> Option<Bytes> {
        match self {
            ValueDeletable::Value(v) => Some(v.clone()),
            ValueDeletable::Merge(_) => {
                unimplemented!("MergeOperator is not yet fully implemented")
            }
            ValueDeletable::Tombstone => None,
        }
    }

    pub fn is_tombstone(&self) -> bool {
        matches!(self, ValueDeletable::Tombstone)
    }
}
