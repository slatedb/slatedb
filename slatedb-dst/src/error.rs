use slatedb::Error;
use thiserror::Error as ThisError;

#[derive(Debug, ThisError)]
pub(crate) enum DstError {
    #[error("failed to perform an operation on the SQLite state database")]
    SQLiteStateError(#[from] rusqlite::Error),
}

impl From<DstError> for Error {
    fn from(err: DstError) -> Self {
        let msg = err.to_string();
        match err {
            DstError::SQLiteStateError(e) => Error::operation(msg).with_source(Box::new(e)),
        }
    }
}
