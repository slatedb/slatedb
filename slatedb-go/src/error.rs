use slatedb::Error as SlateError;
use std::ffi::{CStr, CString};
use std::os::raw::c_char;

// Error codes that will be exposed to C/Go
#[repr(C)]
#[derive(Debug, Copy, Clone, PartialEq)]
pub enum CSdbError {
    Success = 0,
    InvalidArgument = 1,
    NotFound = 2,
    AlreadyExists = 3,
    IOError = 4,
    InternalError = 5,
    NullPointer = 6,
    InvalidHandle = 7,
}

// Result type for returning both error codes and messages
#[repr(C)]
pub struct CSdbResult {
    pub error: CSdbError,
    pub message: *mut c_char,
}

// Helper functions for error handling
pub fn create_error_result(error: CSdbError, message: &str) -> CSdbResult {
    let c_message =
        CString::new(message).unwrap_or_else(|_| CString::new("Invalid UTF-8").unwrap());
    CSdbResult {
        error,
        message: c_message.into_raw(),
    }
}

pub fn create_success_result() -> CSdbResult {
    CSdbResult {
        error: CSdbError::Success,
        message: std::ptr::null_mut(),
    }
}

pub fn safe_str_from_ptr(ptr: *const c_char) -> Result<&'static str, CSdbError> {
    if ptr.is_null() {
        return Err(CSdbError::NullPointer);
    }

    unsafe {
        CStr::from_ptr(ptr)
            .to_str()
            .map_err(|_| CSdbError::InvalidArgument)
    }
}

pub fn slate_error_to_code(error: &SlateError) -> CSdbError {
    // Use string matching since we can't pattern match on the exact variants
    let error_str = format!("{:?}", error);
    if error_str.contains("NotFound") {
        CSdbError::NotFound
    } else if error_str.contains("AlreadyExists") {
        CSdbError::AlreadyExists
    } else if error_str.contains("InvalidRequest") {
        CSdbError::InvalidArgument
    } else if error_str.contains("IO") {
        CSdbError::IOError
    } else {
        CSdbError::InternalError
    }
}
