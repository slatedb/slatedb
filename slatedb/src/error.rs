use object_store::path::Path;
use std::any::Any;
use std::ops::Bound;
use std::sync::Mutex;
use std::time::Duration;
use std::{path::PathBuf, sync::Arc};
use thiserror::Error as ThisError;
use uuid::Uuid;

use crate::bytes_range::BytesRange;
use crate::merge_operator::MergeOperatorError;

#[non_exhaustive]
#[derive(Clone, Debug, ThisError)]
pub(crate) enum SlateDBError {
    #[error("io error")]
    IoError(#[from] Arc<std::io::Error>),

    #[error("checksum mismatch")]
    ChecksumMismatch,

    #[error("empty SSTable")]
    EmptySSTable,

    #[error("empty block metadata")]
    EmptyBlockMeta,

    #[error("empty block")]
    EmptyBlock,

    #[error("empty manifest")]
    EmptyManifest,

    #[error("object store error")]
    ObjectStoreError(#[from] Arc<object_store::Error>),

    #[error("manifest file already exists")]
    ManifestVersionExists,

    #[error("failed to find manifest with id. id=`{0}`")]
    ManifestMissing(u64),

    #[error("failed to find latest manifest")]
    LatestManifestMissing,

    #[error("invalid deletion")]
    InvalidDeletion,

    #[error("invalid sst error")]
    InvalidFlatbuffer(#[from] flatbuffers::InvalidFlatbuffer),

    #[error("invalid DB state error")]
    InvalidDBState,

    #[error("unsupported operation. operation=`{0}`")]
    Unsupported(String),

    #[error("invalid compaction")]
    InvalidCompaction,

    #[error("compaction executor failed")]
    CompactionExecutorFailed,

    #[error(
        "invalid clock tick, must be monotonic. last_tick=`{last_tick}`, next_tick=`{next_tick}`"
    )]
    InvalidClockTick { last_tick: i64, next_tick: i64 },

    #[error("detected newer DB client")]
    Fenced,

    #[error("invalid cache part size bytes, it must be multiple of 1024 and greater than 0")]
    InvalidCachePartSize,

    #[error("invalid compression codec")]
    InvalidCompressionCodec,

    #[cfg(any(
        feature = "snappy",
        feature = "zlib",
        feature = "lz4",
        feature = "zstd"
    ))]
    #[error("error decompressing block")]
    BlockDecompressionError,

    #[cfg(any(feature = "snappy", feature = "zlib", feature = "zstd"))]
    #[error("error compressing block")]
    BlockCompressionError,

    #[error("Invalid RowFlags. #{message}. encoded_bits=`{encoded_bits:#b}`, known_bits=`{known_bits:#b}`"
    )]
    InvalidRowFlags {
        encoded_bits: u8,
        known_bits: u8,
        message: String,
    },

    #[error("read channel error")]
    ReadChannelError(#[from] tokio::sync::oneshot::error::RecvError),

    #[error("iterator invalidated after unexpected error")]
    InvalidatedIterator(#[from] Box<SlateDBError>),

    #[error("background task panic'd")]
    // we need to wrap the panic args in an Arc so SlateDbError is Clone
    // we need to wrap the panic args in a mutex so that SlateDbError is Sync
    BackgroundTaskPanic(Arc<Mutex<Box<dyn Any + Send>>>),

    #[error("background task shutdown")]
    BackgroundTaskShutdown,

    #[error("merge operator error")]
    MergeOperatorError(#[from] MergeOperatorError),

    #[error("checkpoint missing. checkpoint_id=`{0}`")]
    CheckpointMissing(Uuid),

    #[error(
        "byte format version mismatch. expected_version=`{expected_version}`, actual_version=`{actual_version}`"
    )]
    InvalidVersion {
        expected_version: u16,
        actual_version: u16,
    },

    #[error("foyer cache reading error")]
    #[cfg(feature = "foyer")]
    FoyerCacheReadingError(#[from] Arc<anyhow::Error>),

    #[error("operation timed out. operation=`{op}`")]
    Timeout { op: &'static str, backoff: Duration },

    #[error("wal buffer already started")]
    WalBufferAlreadyStarted,

    #[error("cannot seek to a key outside the iterator range. key=`{key:?}`, range=`{range:?}`")]
    SeekKeyOutOfRange { key: Vec<u8>, range: BytesRange },

    #[error("cannot seek to a key less than the last returned key")]
    SeekKeyLessThanLastReturnedKey,

    #[error(
        "parent path must be different from the clone's path. parent_path=`{0}`, clone_path=`{0}`"
    )]
    IdenticalClonePaths(Path),

    #[error("invalid checkpoint lifetime. lifetime=`{0:?}`")]
    InvalidCheckpointLifetime(Duration),

    #[error("invalid manifest poll interval. interval=`{0:?}`")]
    InvalidManifestPollInterval(Duration),

    #[error("checkpoint lifetime must be at least double the manifest poll interval. lifetime=`{lifetime:?}`, interval=`{interval:?}`")]
    CheckpointLifetimeTooShort {
        lifetime: Duration,
        interval: Duration,
    },

    #[error("invalid sst batch size. size=`{0}`")]
    InvalidSSTBatchSize(usize),

    #[error("cannot seek to a key outside the iterator range. key=`{key:?}`, start_key=`{start_key:?}`, end_key=`{end_key:?}`")]
    SeekKeyOutOfKeyRange {
        key: Vec<u8>,
        start_key: Bound<Vec<u8>>,
        end_key: Bound<Vec<u8>>,
    },

    #[error("the cloned database is not attached to any external database")]
    CloneExternalDbMissing,

    #[error("the cloned database is not attached to external database with a valid checkpoint. path=`{path}`, checkpoint_id=`{checkpoint_id:?}`")]
    CloneIncorrectExternalDbCheckpoint {
        path: String,
        checkpoint_id: Option<Uuid>,
    },

    #[error("the final checkpoint for the cloned database no longer exists in the manifest. path=`{path}`, checkpoint_id=`{checkpoint_id}`")]
    CloneIncorrectFinalCheckpoint { path: String, checkpoint_id: Uuid },

    #[error("unknown configuration file format. path=`{0}`")]
    UnknownConfigurationFormat(PathBuf),

    #[error("invalid configuration format")]
    InvalidConfigurationFormat(#[from] Box<figment::Error>),

    #[error("attempted a WAL operation when the WAL is disabled")]
    WalDisabled,

    #[error("invalid object store URL. url=`{0}`")]
    InvalidObjectStoreURL(String, #[source] url::ParseError),
}

impl From<std::io::Error> for SlateDBError {
    fn from(value: std::io::Error) -> Self {
        Self::IoError(Arc::new(value))
    }
}

impl From<object_store::Error> for SlateDBError {
    fn from(value: object_store::Error) -> Self {
        Self::ObjectStoreError(Arc::new(value))
    }
}

type BoxError = Box<dyn std::error::Error + Send + Sync>;

/// Represents the kind of public errors that can be returned to the user.
#[derive(Debug)]
pub enum ErrorKind {
    /// The database attempted to use an invalid configuration.
    Configuration,

    /// The database attempted an invalid operation or an operation with an
    /// invalid parameter (including misconfiguration).
    Operation,

    /// Unexpected internal error. This error is fatal (i.e. the database must be closed).
    System,

    /// Invalid persistent state (e.g. corrupted data files). The state must
    /// be repaired before the database can be restarted.
    PersistentState,

    /// Failed access database resources (e.g. remote storage) due to some
    /// kind of authentication or authorization error. The operation can be
    /// retried after the permission issue is resolved.
    Permission,

    /// The operation failed due to a transient error (such as IO unavailability).
    /// The operation can be retried after backing off.
    Transient(Duration),
}

impl std::fmt::Display for ErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ErrorKind::Configuration => write!(f, "Configuration error"),
            ErrorKind::Operation => write!(f, "Operation error"),
            ErrorKind::System => write!(f, "System error"),
            ErrorKind::PersistentState => write!(f, "Persistent state error"),
            ErrorKind::Permission => write!(f, "Permission error"),
            ErrorKind::Transient(backoff) => {
                write!(f, "Transient error (retry after: {backoff:?})")
            }
        }
    }
}

/// Represents a public error that can be returned to the user.
#[derive(Debug)]
pub struct Error {
    msg: String,
    kind: ErrorKind,
    source: Option<BoxError>,
}

impl std::fmt::Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.kind, self.msg)?;
        if let Some(source) = self.source.as_ref() {
            write!(f, " ({source})")?;
        }
        Ok(())
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.source
            .as_ref()
            .map(|e| e.as_ref() as &(dyn std::error::Error + 'static))
    }
}

impl Error {
    /// Creates a new public configuration error.
    pub fn configuration(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Configuration,
            source: None,
        }
    }

    /// Creates a new public operation error.
    pub fn operation(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Operation,
            source: None,
        }
    }

    /// Creates a new public system error.
    pub fn system(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::System,
            source: None,
        }
    }

    /// Creates a new public persistent state error.
    pub fn persistent_state(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::PersistentState,
            source: None,
        }
    }

    /// Creates a new public permission error.
    pub fn permission(msg: String) -> Self {
        Self {
            msg,
            kind: ErrorKind::Permission,
            source: None,
        }
    }

    /// Creates a new public transient error.
    pub fn transient(msg: String, backoff: Duration) -> Self {
        Self {
            msg,
            kind: ErrorKind::Transient(backoff),
            source: None,
        }
    }

    /// Adds a source to the error.
    pub fn with_source(mut self, source: BoxError) -> Self {
        self.source = Some(source);
        self
    }
}

impl From<SlateDBError> for Error {
    fn from(err: SlateDBError) -> Self {
        let msg = err.to_string();
        match err {
            SlateDBError::IoError(err) => Error::system(msg).with_source(Box::new(err)),
            SlateDBError::ObjectStoreError(err) => Error::operation(msg).with_source(Box::new(err)),
            SlateDBError::InvalidFlatbuffer(err) => {
                Error::persistent_state(msg).with_source(Box::new(err))
            }
            SlateDBError::InvalidDeletion => Error::operation(msg),
            SlateDBError::InvalidDBState => Error::persistent_state(msg),
            SlateDBError::InvalidCompaction => Error::operation(msg),
            SlateDBError::CompactionExecutorFailed => Error::system(msg),
            SlateDBError::InvalidClockTick { .. } => Error::operation(msg),
            SlateDBError::Fenced => Error::permission(msg),
            SlateDBError::InvalidCachePartSize => Error::operation(msg),
            SlateDBError::InvalidCompressionCodec => Error::operation(msg),
            #[cfg(any(
                feature = "snappy",
                feature = "zlib",
                feature = "lz4",
                feature = "zstd"
            ))]
            SlateDBError::BlockDecompressionError => Error::operation(msg),
            #[cfg(any(feature = "snappy", feature = "zlib", feature = "zstd"))]
            SlateDBError::BlockCompressionError => Error::operation(msg),
            SlateDBError::InvalidRowFlags { .. } => Error::operation(msg),
            SlateDBError::ReadChannelError(err) => Error::system(msg).with_source(Box::new(err)),
            SlateDBError::InvalidatedIterator(err) => {
                Error::operation(msg).with_source(Box::new(err))
            }
            SlateDBError::BackgroundTaskPanic(err) => {
                Error::system(msg).with_source(Box::new(PanicError(err)))
            }
            SlateDBError::BackgroundTaskShutdown => Error::system(msg),
            SlateDBError::MergeOperatorError(err) => {
                Error::operation(msg).with_source(Box::new(err))
            }
            SlateDBError::CheckpointMissing(_) => Error::persistent_state(msg),
            SlateDBError::InvalidVersion { .. } => Error::persistent_state(msg),
            #[cfg(feature = "foyer")]
            SlateDBError::FoyerCacheReadingError(err) => {
                Error::persistent_state(msg).with_source(Box::new(AnyhowError(err)))
            }
            SlateDBError::Timeout { backoff, .. } => Error::transient(msg, backoff),
            SlateDBError::WalBufferAlreadyStarted => Error::system(msg),
            SlateDBError::ManifestMissing(_) => Error::persistent_state(msg),
            SlateDBError::LatestManifestMissing => Error::persistent_state(msg),
            SlateDBError::ManifestVersionExists => Error::persistent_state(msg),
            SlateDBError::EmptyManifest => Error::persistent_state(msg),
            SlateDBError::EmptyBlock => Error::persistent_state(msg),
            SlateDBError::EmptyBlockMeta => Error::persistent_state(msg),
            SlateDBError::EmptySSTable => Error::persistent_state(msg),
            SlateDBError::Unsupported(_) => Error::operation(msg),
            SlateDBError::ChecksumMismatch => Error::persistent_state(msg),
            SlateDBError::SeekKeyOutOfRange { .. } => Error::operation(msg),
            SlateDBError::SeekKeyLessThanLastReturnedKey => Error::operation(msg),
            SlateDBError::IdenticalClonePaths { .. } => Error::operation(msg),
            SlateDBError::InvalidCheckpointLifetime(_) => Error::operation(msg),
            SlateDBError::InvalidManifestPollInterval(_) => Error::operation(msg),
            SlateDBError::CheckpointLifetimeTooShort { .. } => Error::operation(msg),
            SlateDBError::InvalidSSTBatchSize(_) => Error::operation(msg),
            SlateDBError::SeekKeyOutOfKeyRange { .. } => Error::operation(msg),
            SlateDBError::CloneExternalDbMissing => Error::persistent_state(msg),
            SlateDBError::CloneIncorrectExternalDbCheckpoint { .. } => Error::persistent_state(msg),
            SlateDBError::CloneIncorrectFinalCheckpoint { .. } => Error::persistent_state(msg),
            SlateDBError::UnknownConfigurationFormat(_) => Error::configuration(msg),
            SlateDBError::InvalidConfigurationFormat(err) => {
                Error::configuration(msg).with_source(Box::new(err))
            }
            SlateDBError::WalDisabled => Error::operation(msg),
            SlateDBError::InvalidObjectStoreURL(_, err) => {
                Error::configuration(msg).with_source(Box::new(err))
            }
        }
    }
}

#[derive(Debug)]
struct PanicError(Arc<Mutex<Box<dyn Any + Send>>>);
impl std::error::Error for PanicError {}
impl std::fmt::Display for PanicError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let guard = self.0.lock().expect("Failed to lock panic error");
        if let Some(err) = guard.downcast_ref::<SlateDBError>() {
            write!(f, "{err}")?;
        } else if let Some(err) = guard.downcast_ref::<Box<dyn std::error::Error>>() {
            write!(f, "{err}")?;
        } else if let Some(err) = guard.downcast_ref::<String>() {
            write!(f, "{err}")?;
        } else {
            write!(f, "irrecoverable panic")?;
        }
        Ok(())
    }
}

#[derive(Debug)]
#[cfg(feature = "foyer")]
struct AnyhowError(Arc<anyhow::Error>);
#[cfg(feature = "foyer")]
impl std::error::Error for AnyhowError {}
#[cfg(feature = "foyer")]
impl std::fmt::Display for AnyhowError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
