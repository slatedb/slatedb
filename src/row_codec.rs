use crate::db_state::RowAttribute;
use crate::error::SlateDBError;
use crate::types::{KeyValueDeletable, RowAttributes, ValueDeletable};
use bitflags::bitflags;
use bytes::{Buf, BufMut, Bytes, BytesMut};

bitflags! {
    pub(crate) struct RowFlags: u8 {
        const Tombstone = 0b00000001;
    }
}

/// Encodes key and value using the binary codec for SlateDB row representation
/// using the `v0` encoding scheme.
///
/// The `v0` codec for the key is (for non-tombstones):
///
/// ```txt
///  |--------------------------------------------------------------------------|
///  |       u16      |      u16       |  var        | var  |    u32    |  var  |
///  |----------------|----------------|-------------|------|-----------|-------|
///  | key_prefix_len | key_suffix_len |  key_suffix | meta | value_len | value |
///  |--------------------------------------------------------------------------|
/// ```
///
/// And for tombstones (with `meta[row_flags] & 0x01 == 1`):
///  ```txt
///  |------------------------------------------------------|
///  |       u16      |      u16       |  var        | var  |
///  |----------------|----------------|-------------|------|
///  | key_prefix_len | key_suffix_len |  key_suffix | meta |
///  |------------------------------------------------------|
///  ```
///
/// | Field            | Type | Description                                   |
/// |------------------|------|-----------------------------------------------|
/// | `key_prefix_len` | `u16` | Length of the key prefix                     |
/// | `key_suffix_len` | `u16` | Length of the key suffix                     |
/// | `key_suffix`     | `var` | Suffix of the key                            |
/// | `meta`           | `var` | Metadata                                     |
/// | `value_len`      | `u32` | Length of the value                          |
/// | `value`          | `var` | Value bytes                                  |
pub(crate) fn encode_row_v0(
    data: &mut Vec<u8>,
    key_prefix_len: usize,
    key_suffix: &[u8],
    value: Option<&[u8]>,
    row_attributes: &Vec<RowAttribute>,
    timestamp: Option<i64>,
) {
    data.put_u16(key_prefix_len as u16);
    data.put_u16(key_suffix.len() as u16);
    data.put(key_suffix);

    data.put(encode_meta(row_attributes, value.is_none(), timestamp));

    if let Some(value) = value {
        data.put_u32(value.len() as u32);
        data.put(value);
    }
}

fn encode_meta(
    row_attributes: &Vec<RowAttribute>,
    is_tombstone: bool,
    timestamp: Option<i64>,
) -> Bytes {
    let mut meta = BytesMut::new();

    for attr in row_attributes {
        match attr {
            RowAttribute::Flags => meta.put_u8(encode_row_flags(is_tombstone)),
            RowAttribute::Timestamp => meta.put_i64(timestamp.expect("Timestamp RowAttribute was enabled for SST but attempted to insert a row with no timestamp.")),
        }
    }

    meta.freeze()
}

fn encode_row_flags(is_tombstone: bool) -> u8 {
    let mut flags = RowFlags::empty();

    if is_tombstone {
        flags |= RowFlags::Tombstone;
    }

    flags.bits()
}

/// Decodes a row based on the encoding specified in [`encode_row_v0`]. If the
/// `data` passed in was not encoded using `encode_row_v0`, or if there are any
/// unknown row attributes, this method will return an error.
pub(crate) fn decode_row_v0(
    first_key: &Bytes,
    row_attributes: &Vec<RowAttribute>,
    data: &mut Bytes,
) -> Result<KeyValueDeletable, SlateDBError> {
    let key_prefix_len = data.get_u16() as usize;
    let key_suffix_len = data.get_u16() as usize;
    let key_suffix = data.slice(..key_suffix_len);
    data.advance(key_suffix_len);

    let meta = decode_meta(row_attributes, data)?;
    let value = if !(meta.flags & RowFlags::Tombstone).is_empty() {
        ValueDeletable::Tombstone
    } else {
        let value_len = data.get_u32() as usize;
        ValueDeletable::Value(data.slice(..value_len))
    };

    let mut key = BytesMut::with_capacity(key_prefix_len + key_suffix_len);
    key.extend_from_slice(&first_key[..key_prefix_len]);
    key.extend_from_slice(&key_suffix);

    Ok(KeyValueDeletable {
        key: key.into(),
        value,
        attributes: RowAttributes { ts: meta.timestamp },
    })
}

struct RowMetadata {
    flags: RowFlags,
    timestamp: Option<i64>,
}

fn decode_meta(
    row_attributes: &Vec<RowAttribute>,
    data: &mut Bytes,
) -> Result<RowMetadata, SlateDBError> {
    let mut meta = RowMetadata {
        flags: RowFlags::empty(),
        timestamp: None,
    };
    for attr in row_attributes {
        match attr {
            RowAttribute::Flags => match decode_row_flags(data.get_u8()) {
                Ok(flags) => {
                    meta.flags = flags;
                }
                Err(e) => return Err(e),
            },
            RowAttribute::Timestamp => meta.timestamp = Some(data.get_i64()),
        }
    }
    Ok(meta)
}

fn decode_row_flags(flags: u8) -> Result<RowFlags, SlateDBError> {
    match RowFlags::from_bits(flags) {
        None => Err(SlateDBError::InvalidRowFlags),
        Some(flags) => Ok(flags),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::db_state::RowAttribute;
    use crate::types::ValueDeletable;

    #[test]
    fn test_encode_decode_normal_row() {
        let mut encoded_data = Vec::new();
        let key_prefix_len = 3;
        let key_suffix = b"key";
        let value = Some(b"value".as_slice());
        let row_attributes = vec![RowAttribute::Flags, RowAttribute::Timestamp];

        // Encode the row
        encode_row_v0(
            &mut encoded_data,
            key_prefix_len,
            key_suffix,
            value,
            &row_attributes,
            Some(1),
        );

        let first_key = Bytes::from(b"prefixdata".as_ref());
        let mut data = Bytes::from(encoded_data);
        let decoded =
            decode_row_v0(&first_key, &row_attributes, &mut data).expect("Decoding failed");

        // Expected key: first_key[..3] + "key" = "prekey"
        let expected_key = Bytes::from(b"prekey" as &[u8]);
        let expected_value = ValueDeletable::Value(Bytes::from(b"value" as &[u8]));

        assert_eq!(decoded.key, expected_key);
        assert_eq!(decoded.value, expected_value);
        assert_eq!(decoded.attributes.ts, Some(1));
    }

    #[test]
    fn test_encode_decode_row_with_disabled_row_attr_flag() {
        let mut encoded_data = Vec::new();
        let key_prefix_len = 0;
        let key_suffix = b"";
        let value = Some(b"value".as_slice());
        let row_attributes = vec![RowAttribute::Flags];

        // Encode the row
        encode_row_v0(
            &mut encoded_data,
            key_prefix_len,
            key_suffix,
            value,
            &row_attributes,
            Some(1), // pass in a timestamp, but it should not be encoded because flag is disabled
        );

        let first_key = Bytes::from(b"".as_ref());
        let mut data = Bytes::from(encoded_data);
        let decoded =
            decode_row_v0(&first_key, &row_attributes, &mut data).expect("Decoding failed");

        assert_eq!(decoded.attributes.ts, None);
    }

    #[test]
    fn test_encode_decode_tombstone_row() {
        let mut encoded_data = Vec::new();
        let key_prefix_len = 4;
        let key_suffix = b"tomb";
        let value = None; // Indicates tombstone
        let row_attributes = vec![RowAttribute::Flags];

        // Encode the tombstone row
        encode_row_v0(
            &mut encoded_data,
            key_prefix_len,
            key_suffix,
            value,
            &row_attributes,
            Some(1),
        );

        let first_key = Bytes::from(b"deadbeefdata".as_ref());
        let mut data = Bytes::from(encoded_data);
        let decoded =
            decode_row_v0(&first_key, &row_attributes, &mut data).expect("Decoding failed");

        // Expected key: first_key[..4] + "tomb" = "deadtomb"
        let expected_key = Bytes::from(b"deadtomb" as &[u8]);
        let expected_value = ValueDeletable::Tombstone;

        assert_eq!(decoded.key, expected_key);
        assert_eq!(decoded.value, expected_value);
    }

    #[test]
    fn test_decode_invalid_flags() {
        let mut encoded_data = Vec::new();
        let key_prefix_len = 3;
        let key_suffix = b"bad".as_slice();
        let row_attributes = vec![RowAttribute::Flags];

        // Manually encode invalid flags
        encoded_data.put_u16(key_prefix_len as u16);
        encoded_data.put_u16(key_suffix.len() as u16);
        encoded_data.put(key_suffix);
        encoded_data.put_u8(0xFF); // Invalid flag
        encoded_data.put_u32(4);
        encoded_data.put(b"data".as_slice());

        let first_key = Bytes::from(b"prefixdata".as_ref());
        let mut data = Bytes::from(encoded_data);

        // Attempt to decode the row
        let result = decode_row_v0(&first_key, &row_attributes, &mut data);

        assert!(result.is_err());
        match result {
            Err(SlateDBError::InvalidRowFlags) => (),
            _ => panic!("Expected InvalidRowFlags"),
        }
    }

    #[test]
    fn test_encode_decode_empty_key_suffix() {
        let mut encoded_data = Vec::new();
        let key_prefix_len = 4;
        let key_suffix = b""; // Empty key suffix
        let value = Some(b"value".as_slice());
        let row_attributes = vec![RowAttribute::Flags];

        // Encode the row
        encode_row_v0(
            &mut encoded_data,
            key_prefix_len,
            key_suffix,
            value,
            &row_attributes,
            Some(1),
        );

        let first_key = Bytes::from(b"keyprefixdata".as_slice());
        let mut data = Bytes::from(encoded_data);
        let decoded =
            decode_row_v0(&first_key, &row_attributes, &mut data).expect("Decoding failed");

        // Expected key: first_key[..4] + "" = "keyp"
        let expected_key = Bytes::from(b"keyp" as &[u8]);
        let expected_value = ValueDeletable::Value(Bytes::from(b"value" as &[u8]));

        assert_eq!(decoded.key, expected_key);
        assert_eq!(decoded.value, expected_value);
    }
}
