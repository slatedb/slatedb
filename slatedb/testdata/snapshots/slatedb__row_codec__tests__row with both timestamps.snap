---
source: src/row_codec.rs
expression: output
snapshot_kind: text
---
(
    "row with both timestamps",
    "\0\u{5}\0\u{4}both\0\0\0\0\0\0\0d\u{6}\0\0\0\u{2}L�\u{16}�\0\0\0\0I�\u{2}�\0\0\0\u{5}value",
    SstRowEntry {
        key_prefix_len: 5,
        key_suffix: b"both",
        seq: 100,
        expire_ts: Some(
            9876543210,
        ),
        create_ts: Some(
            1234567890,
        ),
        value: Value(
            b"value",
        ),
    },
    b"test_both",
)
